src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -1050,197 +1036,265 @@ void ImageReplayer<I>::shut_down_journal_replay(bool cancel_ops)
 }
 
 template <typename I>
-void ImageReplayer<I>::reschedule_update_status_task(int new_interval)
-{
-  Mutex::Locker locker(m_threads->timer_lock);
+void ImageReplayer<I>::reschedule_update_status_task(int new_interval) {
+  dout(20) << dendl;
+  m_lock.Lock();
+  m_threads->timer_lock.Lock();
 
+  bool canceled_task = false;
   if (m_update_status_task) {
-    m_threads->timer->cancel_event(m_update_status_task);
+    canceled_task = m_threads->timer->cancel_event(m_update_status_task);
     m_update_status_task = nullptr;
   }
 
   if (new_interval > 0) {
     m_update_status_interval = new_interval;
   }
 
-  if (new_interval < 0) {
-    return;
-  }
+  bool restarting = (new_interval == 0 || canceled_task);
+  if (new_interval >= 0 && is_running_() &&
+      start_mirror_image_status_update(false, restarting)) {
+    m_update_status_task = new FunctionContext(
+      [this](int r) {
+        assert(m_threads->timer_lock.is_locked());
+        m_update_status_task = nullptr;
 
-  m_update_status_task = new FunctionContext(
-    [this](int r) {
-      start_update_status_task();
-    });
+        queue_mirror_image_status_update(boost::none);
+      });
+    m_threads->timer->add_event_after(m_update_status_interval,
+                                      m_update_status_task);
+  }
+  m_threads->timer_lock.Unlock();
+  m_lock.Unlock();
 
-  m_threads->timer->add_event_after(m_update_status_interval,
-				    m_update_status_task);
+  if (canceled_task) {
+    dout(20) << "canceled task" << dendl;
+    finish_mirror_image_status_update();
+  }
 }
 
 template <typename I>
-void ImageReplayer<I>::start_update_status_task()
-{
-  FunctionContext *ctx = new FunctionContext(
-    [this](int r) {
-      {
-	Mutex::Locker locker(m_threads->timer_lock);
-	m_update_status_task = nullptr;
-      }
-      update_mirror_image_status();
-    });
-  m_threads->work_queue->queue(ctx, 0);
+void ImageReplayer<I>::handle_stop(int r, Context *on_start, Context *on_stop) {
+  reschedule_update_status_task(-1);
+
+  {
+    Mutex::Locker locker(m_lock);
+
+    // if status updates are in-flight, wait for them to complete
+    // before proceeding
+    if (m_in_flight_status_updates > 0) {
+      dout(20) << "waiting for in-flight status update" << dendl;
+      assert(m_on_update_status_finish == nullptr);
+      m_on_update_status_finish = new FunctionContext(
+        [this, r, on_start, on_stop](int r) {
+          handle_stop(r, on_start, on_stop);
+        });
+      return;
+    }
+
+    m_stop_requested = false;
+    m_state = STATE_STOPPED;
+    m_state_desc.clear();
+    m_last_r = 0;
+  }
+  dout(20) << "stop complete" << dendl;
+
+  m_local_ioctx.close();
+  m_remote_ioctx.close();
+
+  if (on_start != nullptr) {
+    dout(20) << "on start finish complete, r=" << r << dendl;
+    on_start->complete(r);
+  }
+  if (on_stop != nullptr) {
+    dout(20) << "on stop finish complete, r=" << r << dendl;
+    on_stop->complete(on_start == nullptr ? r : 0);
```

Comment:
minor: it might still be confusing to see in the debug log message `r`, while complete is actually called with 0.